### PR TITLE
Record user tours views

### PR DESCRIPTION
**Pull request by nathanstitt**
Saves viewing statistics for users viewing tours.

Is pretty bare-bones but should give us a a place to start from.

The "Tour" record is auto-created so the FE can supply an arbitrary id like "teacher-dashboard"

### CODE REVIEW

File: app/subsystems/user/models/tour_view.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+module User
+  module Models
+    class TourView < Tutor::SubSystems::BaseModel
+
+      belongs_to :tour
```
Comment (Dantemss) [inline app/subsystems/user/models/tour_view.rb]: Could add `inverse_of: :tour_views` here after defining it in `User::Models::Tour`.

File: app/subsystems/user/models/tour_view.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+      belongs_to :profile, -> { with_deleted }, inverse_of: :tour_views
+
+    end
```
Comment (Dantemss) [inline app/subsystems/user/models/tour_view.rb]: Could add presence validations for tour and profile and a uniqueness validation for profile scoped to `:user_tour_id`.

File: app/subsystems/user/record_tour_view.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
+module User
+
+
+  class RecordTourView
+    lev_routine express_output: :view
+
+    protected
+
+    def exec(user:, tour_identifier: )
```
Comment (Dantemss) [inline app/subsystems/user/record_tour_view.rb]: 2 find_or_create race conditions here but not sure if we can do much until postgres 9.5...

File: app/subsystems/user/models/tour_view.rb (hunk 2 above)
Comment (Dantemss) [inline app/subsystems/user/models/tour_view.rb]: And validate presence of `view_count`.

File: db/migrate/20160923180438_create_user_tours.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+  def change
+    create_table :user_tours do |t|
+      t.text :identifier, null: false, index: { unique: true }
+      t.timestamps null: false
+    end
+
+    create_table :user_tour_views do |t|
+      t.integer :view_count, null: false, default: 0
+      t.belongs_to :user_profile, foreign_key: true, null: false, index: { unique: false }
```
Comment (Dantemss) [inline db/migrate/20160923180438_create_user_tours.rb]: You probably want to get rid of this index and add a unique index on `[:user_profile_id, :user_tour_id]`

File: db/migrate/20160923180438_create_user_tours.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+      t.belongs_to :user_tour,    foreign_key: true, null: false, index: { unique: false }
```
Comment (Dantemss) [inline db/migrate/20160923180438_create_user_tours.rb]: I think you can just say `index: true` here

File: spec/factories/tours.rb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+FactoryGirl.define do
+  factory :tour do
+    
+  end
+
+end
```
Comment (Dantemss) [inline spec/factories/tours.rb]: I think this empty factory can be removed (seems like you moved the model into the User subsystem).

**Comment by nathanstitt** [comment]
Good points all around @Dantemss 😁

**Pull request merged**